### PR TITLE
 Core/Group: implement automatic party/raid leader change when the leader has been offline for two minutes

### PR DESCRIPTION
**Pull request by Wyrserth**
**Changes proposed:**

Very rough draft.

Added #23396 to make it easier to test without having to pull code from the other PR.

**Target branch(es):** 3.3.5/master

- [x] 3.3.5

**Tests performed:** tested, works.

**Known issues and TODO list:** (add/remove lines as needed)

- [ ] Not implemented yet: changing leader after a server restart. I'm not quite sure on how to tackle that.

### CODE REVIEW

**Comment by Killyana** [comment]
Can you add a check before giving the lead to a player?
Something like:
- If in 5 man group give the lead to the dungeon guide if it exists, connected and different from the lead group
- If raid group give the lead to the firs assist online if it exists.

**Comment by Wyrserth** [comment]
Added priority for raid assistants. In 5-man dungeon finder the dungeon guide is the leader, so it should be fine already :P

**Comment by persephone** [comment]
You are selecting the first player in the group as a new leader, regardless of if it found the main assistant.

As you are overwriting the newPlayer pointer in the second loop no matter what happends in the first

File: src/server/game/Groups/Group.cpp [hunk 1]
```diff
@@ -2554,3 +2600,14 @@ void Group::ToggleGroupMemberFlag(member_witerator slot, uint8 flag, bool apply)
...
+
+void Group::StartLeaderOfflineTimer()
+{
+    m_isLeaderOffline = true;
+}
+
+void Group::StopLeaderOfflineTimer()
+{
+    m_isLeaderOffline = false;
+    m_leaderOfflineTimer.Reset(2 * MINUTE * IN_MILLISECONDS);
```
Comment (jackpoz) [inline src/server/game/Groups/Group.cpp]: I would put the timer in start

**Comment by Wyrserth** [comment]
@VincentVanclef you're right, thanks!

**Comment by jackpoz** [comment]
is this still a draft or is it ready to be merged ?

**Comment by Wyrserth** [comment]
I marked it as draft because I wanted to be sure it got some feedback and maybe testing, I've been testing it for a few days and noticed no issue. I think it can be merged :P

**Comment by jackpoz** [comment]
Thanks for the PR :)

**Pull request merged**